File: src/hotspot/share/gc/z/zPage.cpp
/*
 * Copyright (c) 2015, 2025, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

#include "gc/shared/gc_globals.hpp"
#include "gc/z/zGeneration.inline.hpp"
#include "gc/z/zGlobals.hpp"
#include "gc/z/zPage.inline.hpp"
#include "gc/z/zPageAge.hpp"
#include "gc/z/zRememberedSet.inline.hpp"
#include "utilities/align.hpp"
#include "utilities/debug.hpp"

ZPage::ZPage(ZPageType type, ZPageAge age, const ZVirtualMemory& vmem, ZMultiPartitionTracker* multi_partition_tracker, uint32_t partition_id)
  : _type(type),
    _generation_id(/* set in reset */),
    _age(/* set in reset */),
    _seqnum(/* set in reset */),
    _seqnum_other(/* set in reset */),
    _single_partition_id(partition_id),
    _virtual(vmem),
    _top(to_zoffset_end(start())),
    _livemap(object_max_count()),
    _remembered_set(),
    _multi_partition_tracker(multi_partition_tracker) {
  assert(!_virtual.is_null(), "Should not be null");
  assert((_type == ZPageType::small && size() == ZPageSizeSmall) ||
         (_type == ZPageType::medium && ZPageSizeMediumMin <= size() && size() <= ZPageSizeMediumMax) ||
         (_type == ZPageType::large && is_aligned(size(), ZGranuleSize)),
         "Page type/size mismatch");
  reset(age);

  if (is_old()) {
    remset_alloc();
  }
}

ZPage::ZPage(ZPageType type, ZPageAge age, const ZVirtualMemory& vmem, uint32_t partition_id)
  : ZPage(type, age, vmem, nullptr /* multi_partition_tracker */, partition_id) {}

ZPage::ZPage(ZPageType type, ZPageAge age, const ZVirtualMemory& vmem, ZMultiPartitionTracker* multi_partition_tracker)
  : ZPage(type, age, vmem, multi_partition_tracker, -1u /* partition_id */) {}

ZPage* ZPage::clone_for_promotion() const {
  assert(_age != ZPageAge::old, "must be used for promotion");
  // Only copy type and memory layouts, and also update _top. Let the rest be
  // lazily reconstructed when needed.
  ZPage* const page = new ZPage(_type, ZPageAge::old, _virtual, _multi_partition_tracker, _single_partition_id);
  page->_top = _top;

  return page;
}

ZGeneration* ZPage::generation() {
  return ZGeneration::generation(_generation_id);
}

const ZGeneration* ZPage::generation() const {
  return ZGeneration::generation(_generation_id);
}

void ZPage::reset_seqnum() {
  Atomic::store(&_seqnum, generation()->seqnum());
  Atomic::store(&_seqnum_other, ZGeneration::generation(_generation_id == ZGenerationId::young ? ZGenerationId::old : ZGenerationId::young)->seqnum());
}

void ZPage::remset_alloc() {
  // Remsets should only be allocated/initialized once and only for old pages.
  assert(!_remembered_set.is_initialized(), "Should not be initialized");
  assert(is_old(), "Only old pages need a remset");

  _remembered_set.initialize(size());
}

ZPage* ZPage::reset(ZPageAge age) {
  _age = age;

  _generation_id = age == ZPageAge::old
      ? ZGenerationId::old
      : ZGenerationId::young;

  reset_seqnum();

  return this;
}

void ZPage::reset_livemap() {
  _livemap.reset();
}

void ZPage::reset_top_for_allocation() {
  _top = to_zoffset_end(start());
}

class ZFindBaseOopClosure : public ObjectClosure {
private:
  volatile zpointer* _p;
  oop _result;

public:
  ZFindBaseOopClosure(volatile zpointer* p)
    : _p(p),
      _result(nullptr) {}

  virtual void do_object(oop obj) {
    const uintptr_t p_int = reinterpret_cast<uintptr_t>(_p);
    const uintptr_t base_int = cast_from_oop<uintptr_t>(obj);
    const uintptr_t end_int = base_int + wordSize * obj->size();
    if (p_int >= base_int && p_int < end_int) {
      _result = obj;
    }
  }

  oop result() const { return _result; }
};

bool ZPage::is_remset_cleared_current() const {
  return _remembered_set.is_cleared_current();
}

bool ZPage::is_remset_cleared_previous() const {
  return _remembered_set.is_cleared_previous();
}

void ZPage::verify_remset_cleared_current() const {
  if (ZVerifyRemembered && !is_remset_cleared_current()) {
    fatal_msg(" current remset bits should be cleared");
  }
}

void ZPage::verify_remset_cleared_previous() const {
  if (ZVerifyRemembered && !is_remset_cleared_previous()) {
    fatal_msg(" previous remset bits should be cleared");
  }
}

void ZPage::clear_remset_previous() {
  _remembered_set.clear_previous();
}

void ZPage::swap_remset_bitmaps() {
  _remembered_set.swap_remset_bitmaps();
}

void* ZPage::remset_current() {
  return _remembered_set.current();
}

void ZPage::print_on_msg(outputStream* st, const char* msg) const {
  st->print_cr("%-6s  " PTR_FORMAT " " PTR_FORMAT " " PTR_FORMAT " %s/%-4u %s%s%s%s",
                type_to_string(), untype(start()), untype(top()), untype(end()),
                is_young() ? "Y" : "O",
                seqnum(),
                is_relocatable() ? " Relocatable" : "",
                is_allocating()  ? " Allocating"  : "",
                is_allocating() && msg != nullptr ? " " : "",
                msg != nullptr ? msg : "");
}

void ZPage::print_on(outputStream* st) const {
  print_on_msg(st, nullptr);
}

void ZPage::print() const {
  print_on(tty);
}

void ZPage::verify_live(uint32_t live_objects, size_t live_bytes, bool in_place) const {
  if (!in_place) {
    // In-place relocation has changed the page to allocating
    assert_zpage_mark_state();
  }
  guarantee(live_objects == _livemap.live_objects(), "Invalid number of live objects");
  guarantee(live_bytes == _livemap.live_bytes(), "Invalid number of live bytes");
}

void ZPage::fatal_msg(const char* msg) const {
  stringStream ss;
  print_on_msg(&ss, msg);
  fatal("%s", ss.base());
}

File: src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/gc/z/ZPage.java
/*
 * Copyright (c) 2018, 2019, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 *
 */

package sun.jvm.hotspot.gc.z;

import java.util.ArrayList;
import java.util.List;

import sun.jvm.hotspot.debugger.Address;
import sun.jvm.hotspot.debugger.OopHandle;
import sun.jvm.hotspot.gc.shared.LiveRegionsProvider;
import sun.jvm.hotspot.memory.MemRegion;
import sun.jvm.hotspot.oops.Oop;
import sun.jvm.hotspot.oops.UnknownOopException;
import sun.jvm.hotspot.runtime.VM;
import sun.jvm.hotspot.runtime.VMObject;
import sun.jvm.hotspot.runtime.VMObjectFactory;
import sun.jvm.hotspot.types.AddressField;
import sun.jvm.hotspot.types.CIntegerField;
import sun.jvm.hotspot.types.Type;
import sun.jvm.hotspot.types.TypeDataBase;

public class ZPage extends VMObject implements LiveRegionsProvider {
    private static CIntegerField typeField;
    private static CIntegerField seqnumField;
    private static long virtualFieldOffset;
    private static AddressField topField;
    private static CIntegerField refcountField;
    private static long forwardingFieldOffset;

    static {
        VM.registerVMInitializedObserver((o, d) -> initialize(VM.getVM().getTypeDataBase()));
    }

    static private synchronized void initialize(TypeDataBase db) {
        Type type = db.lookupType("ZPage");

        typeField = type.getCIntegerField("_type");
        seqnumField = type.getCIntegerField("_seqnum");
        virtualFieldOffset = type.getField("_virtual").getOffset();
        topField = type.getAddressField("_top");
        refcountField = type.getCIntegerField("_refcount");
        forwardingFieldOffset = type.getField("_forwarding").getOffset();
    }

    public ZPage(Address addr) {
        super(addr);
    }

    private byte type() {
        return typeField.getJByte(addr);
    }

    private int seqnum() {
        return seqnumField.getJInt(addr);
    }

    private ZVirtualMemory virtual() {
        return VMObjectFactory.newObject(ZVirtualMemory.class, addr.addOffsetTo(virtualFieldOffset));
    }

    private Address top() {
        return topField.getValue(addr);
    }

    private int refcount() {
        // refcount is uint32_t so need to be cautious when using this field.
        return refcountField.getJInt(addr);
    }

    private ZForwardingTable forwarding() {
        return VMObjectFactory.newObject(ZForwardingTable.class, addr.addOffsetTo(forwardingFieldOffset));
    }

    private boolean is_forwarding() {
        return forwarding().table() != null;
    }

    private boolean is_relocatable() {
        return is_active() && seqnum() < ZGlobals.ZGlobalSeqNum();
    }

    private boolean isPageRelocating() {
        assert(is_active());
        // is_forwarding():  Has a (relocation) forwarding table
        // is_relocatable(): Has not been freed yet
        return is_forwarding() && is_relocatable();
    }

    long start() {
        return virtual().start();
    }

    long size() {
        return virtual().end() - virtual().start();
    }

    Address forward_object(Address from) {
        // Lookup address in forwarding table
        long from_offset = ZAddress.offset(from);
        long from_index = (from_offset - start()) >> object_alignment_shift();
        ZForwardingTableEntry entry = forwarding().find(from_index);
        assert(!entry.is_empty());
        assert(entry.from_index() == from_index);

        return ZAddress.good(entry.to_offset());
    }

    Address relocate_object(Address from) {
        // Lookup address in forwarding table
        long from_offset = ZAddress.offset(from);
        long from_index = (from_offset - start()) >> object_alignment_shift();
        ZForwardingTableEntry entry = forwarding().find(from_index);
        if (!entry.is_empty() && entry.from_index() == from_index) {
          return ZAddress.good(entry.to_offset());
        }

        // There's no relocate operation in the SA.
        // Mimic object pinning and return the good view of the from object.
        return ZAddress.good(from);
    }

    long object_alignment_shift() {
        if (type() == ZGlobals.ZPageTypeSmall) {
            return ZGlobals.ZObjectAlignmentSmallShift();
        } else if (type() == ZGlobals.ZPageTypeMedium) {
            return ZGlobals.ZObjectAlignmentMediumShift;
        } else {
            assert(type() == ZGlobals.ZPageTypeLarge);
            return ZGlobals.ZObjectAlignmentLargeShift;
        }
    }

    long objectAlignmentSize() {
        return 1 << object_alignment_shift();
    }

    public boolean is_active() {
        return refcount() != 0;
    }

    private long getObjectSize(Address good) {
        OopHandle handle = good.addOffsetToAsOopHandle(0);
        Oop obj = null;

        try {
           obj = VM.getVM().getObjectHeap().newOop(handle);
        } catch (UnknownOopException exp) {
          throw new RuntimeException(" UnknownOopException  " + exp);
        }

        return VM.getVM().alignUp(obj.getObjectSize(), objectAlignmentSize());
    }

    private void addNotRelocatedRegions(List<MemRegion> regions) {
        MemRegion mr = null;

        // Some objects have already been forwarded to new locations.
        long topValue = top().asLongValue();
        for (long offsetValue = start(); offsetValue < topValue;) {
            Address from = ZAddress.good(ZUtils.longToAddress(offsetValue));

            Address to = relocate_object(from);

            long byteSize;
            try {
                byteSize = getObjectSize(to);
            } catch (Exception e) {
                // Parsing the ZHeap is inherently unsafe
                // when classes have been unloaded. Dead objects
                // might have stale Klass pointers, and there's
                // no way to get the size of the dead object.
                //
                // If possible, run with -XX:-ClassUnloading
                // to ensure that all Klasses are kept alive.
                System.err.println("Unparsable regions found. Skipping: "
                        + from
                        + " "
                        + from.addOffsetTo(topValue - offsetValue));

                // Can't proceed further. Just return the collected regions.
                return;
            }

            if (from.equals(to)) {
                // Not relocated - add region
                if (mr == null) {
                    mr = new MemRegion(from, 0 /* wordSize */);
                    regions.add(mr);
                }

                long wordSize = byteSize / VM.getVM().getBytesPerWord();
                mr.setWordSize(mr.wordSize() + wordSize);
            } else {
                // Forwarded somewhere else, split region.
                mr = null;
            }

            offsetValue += byteSize;
        }
    }

    public List<MemRegion> getLiveRegions() {
        List<MemRegion> res = new ArrayList<>();

        if (isPageRelocating()) {
            addNotRelocatedRegions(res);
        } else {
            Address start = ZAddress.good(ZUtils.longToAddress(start()));

            // Can't convert top() to a "good" address because it might
            // be at the top of the "offset" range, and therefore also
            // looks like one of the color bits. Instead use the "good"
            // address and add the size.
            long size = top().asLongValue() - start();
            Address end = start.addOffsetTo(size);

            res.add(new MemRegion(start, end));
        }

        return res;
    }
}
